### ISSUE TITLE
ffmpeg return "Cannot allocate memory" when I create a socket connection to an unaccessible server

### ISSUE THREAD

**Issue by xxjjss** [sipwise/rtpengine#822]
In recording-daemon/stream.c
When I try to create a socket connection to a server, if the server is unaccessible, when you got errno 113 "No route to host". Even you did nothing, and this server is totally unrelated to recording daemon),  we will get error : 

> Failed to add frame to mixer: failed to get frame from mixer

Which error is reported in recording-daemon/mix.c, line 268.
av_buffersink_get_frame() will return -12, which means  [Cannot allocate memory] 

But if the server is accessible, not matter the connect is created or not, such as when you got errno 111: "Connection refused", this issue will disappear.

If you want to repro this error, add the following two functions into the project. (you may simply add to stream.c)

```
socket_handler_t socket_handler = -1;
void 	try_a_socket_connection(void){
	if (socket_handler == -1) {
		struct sockaddr_in  server_address;
		server_address.sin_addr.s_addr = inet_addr("10.141.9.122");
// Set to an un-exists server, you will repro this issue
// Set to a valid server and set an invalid port, you will get connection refused, but this issue disappear
        server_address.sin_port = htons(5570);
        server_address.sin_family = AF_INET;

	   socket_handler = socket(AF_INET , SOCK_STREAM , 0);  // create a socket handler

        if (socket_handler >=   0 )  {
			if (connect(socket_handler, &server_address, sizeof(struct sockaddr_in)) < 0) {
				ilog(LOG_ERROR," Couldn't create a socket connection errno: %d : %s", errno, strerror(errno));
				close(socket_handler);
				socket_handler = -1;
        	} 
		}

	}
}

void close_scoket(void) {
	if (socket_handler) {
		close(socket_handler);
	}
}
```

And add these two functions call in stream.c, function stream_handler(handler_t *handler) 
```

if (ret == 0) {
		ilog(LOG_INFO, "EOF on stream %s", stream->name);
		stream_close(stream);

		close_scoket();    // <========= Close here

		goto out;
	}
	else if (ret < 0) {
		if (errno == EAGAIN || errno == EINTR || errno == EWOULDBLOCK)
			goto out;
		ilog(LOG_INFO, "Read error on stream %s: %s", stream->name, strerror(errno));
		stream_close(stream);
		goto out;
	}

	// got a packet
	pthread_mutex_unlock(&stream->lock);

	try_a_socket_connection();   // <=========== Connect here

	if (forward_to){
		if (forward_packet(stream->metafile,buf,ret)) // leaves buf intact
			g_atomic_int_inc(&stream->metafile->forward_failed);
		else
			g_atomic_int_inc(&stream->metafile->forward_count);
	}

```

**Comment by xxjjss** [comment]
Has some one looked into this issue? This error is very strange and very interesting.
As you can see from the code, I simply call to create a socket connection, which is totally unrelated to the other part of this project.
But ffmpeg failed as soon as connect() get errorno 113 (No route to host) 
If we get 111 (Connection refused) ffmpeg works.
Even during the recording daemon is processing, if I power off the socket server, the ffmpeg  will start throwing error .

**Comment by rfuchs** [comment]
Unfortunately I'm completely unable to reproduce this. Using your code and a non-local address, all I get is a stalled thread (blocked in `connect()`) that eventually errors out with `Couldn't create a socket connection errno: 110 : Connection timed out`

Using a local address, I get similar behaviour, less stalling and less blocking, and errors `Couldn't create a socket connection errno: 113 : No route to host`

In both cases, recording still worked.

There's a few issues with your code, one being that `socket_handler_t` is not a recognised type (I assume you mean `int`), and the second (more serious one) being that `socket_handler` is an unprotected global variable in a multi threaded environment, which immediately raises concerns about race conditions. The code also makes attempts at closing file descriptor -1.

AFAICT the only side effects to your code are: 1) `errno` being set, 2) the race conditions mentioned, and 3) threads blocking in `connect()`. I would probably start looking into the race conditions, as you can have things happening like one thread closing a file descriptor that another thread is trying to connect, and multiple threads trying to close the same file descriptor, which can interfere with file descriptors opened by other threads.

**Comment by xxjjss** [comment]
Thank you rfuchs. Yes, the above code is copy-pasted from my testing code.
socket_handler_t is typedef int socket_handler_t;  Which is actually a local variable in my code.
In my code, I created a client manager,  in the above steam.c, call this client manager to post the steam.
In the client manager will  copy the above buffer into a thread protected linked list and  Create a new thread and a new semaphore is the stream id is a new one. Then post a semaphore for this thread.
The subthread listen the semaphore and create a new connection on the first semaphore and grab the data from the linked list (repack in a special way) and send it to a server. Also the sub thread will close the connection and quit upon the close signal received.
In this logic, I don't have any racing issue, tested up-to 1000 concurrent calls.
But I encounter this issue when the socket server power off.
The code I posted here is a simple code that I could repro this issue on my side. 
I now was blocking here, why the function av_buffersink_get_frame() will return -12, which means [Cannot allocate memory] ?
I can't see anything in my code could related to ffmpeg.
